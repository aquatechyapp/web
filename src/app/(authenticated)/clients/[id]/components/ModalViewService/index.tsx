'use client';

import {
  CheckCircleIcon,
  ListChecks,
  XCircleIcon,
  Beaker,
  FlaskConical,
  ImageIcon,
  Mail,
  Download,
  MapPin,
  Trash2,
  Eye
} from 'lucide-react';
import Image from 'next/image';

import { Button } from '@/components/ui/button';
import { Dialog, DialogContent, DialogHeader, DialogTitle } from '@/components/ui/dialog';
import { Accordion, AccordionContent, AccordionItem, AccordionTrigger } from '@/components/ui/accordion';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { zipImages } from '@/lib/js-zip';
import { format } from 'date-fns';
import { Service } from '@/ts/interfaces/Service';
import { Pool } from '@/ts/interfaces/Pool';
import DeleteServiceDialog from '../services-datatable/cancel-dialog';
import { useResendServiceEmail } from '@/hooks/react-query/services/useResendServiceEmail';

type Props = {
  service: Service;
  pool: Pool;
  open: boolean;
  setOpen: (open: boolean) => void;
};

export function ModalViewService({ service, pool, open, setOpen }: Props) {
  const resendEmailMutation = useResendServiceEmail();

  const getStatusMessage = (status: string) => {
    switch (status) {
      case 'InProgress':
        return 'Service is in progress.';
      case 'Open':
        return 'Service is still open.';
      case 'Skipped':
        return 'Service was skipped.';
      default:
        return '';
    }
  };

  const handleResendEmail = () => {
    if (service?.id) {
      resendEmailMutation.mutate(
        { serviceId: service.id },
        {
          onSuccess: () => {
            alert('Email resent successfully!');
          },
          onError: () => {
            alert('Failed to resend email. Please try again.');
          }
        }
      );
    }
  };

  // Helper functions for new structured data
  const getStructuredPhotos = () => {
    if (service?.structuredPhotos && service.structuredPhotos.length > 0) {
      return service.structuredPhotos;
    }
    // Fallback to old photos array
    if (service?.photos && service.photos.length > 0) {
      return service.photos.map((photo: string, index: number) => ({
        id: `legacy-${index}`,
        url: photo,
        photoDefinitionId: 'legacy',
        notes: null,
        createdAt: service.completedAt || new Date().toISOString()
      }));
    }
    return [];
  };

  const getPhotosForDownload = () => {
    const structuredPhotos = getStructuredPhotos();
    return structuredPhotos.map((photo: any) => photo.url);
  };

  const hasPhotos = () => {
    const photos = getStructuredPhotos();
    return photos && photos.length > 0;
  };

  const getPhotoGroups = () => {
    if (!service?.photosSnapshot) return [];
    
    return service.photosSnapshot.map((group: any) => {
      const groupPhotos = service?.structuredPhotos?.filter(
        (photo: any) => group.photoDefinitions.some((def: any) => def.id === photo.photoDefinitionId)
      ) || [];
      
      return {
        ...group,
        photos: groupPhotos
      };
    });
  };

  const getReadingGroups = () => {
    if (service?.readingsSnapshot) {
      return service.readingsSnapshot.map((group: any) => ({
        ...group,
        readings: service?.readings?.filter((reading: any) => 
          group.readingDefinitions.some((def: any) => def.id === reading.readingDefinitionId)
        ) || []
      }));
    }
    
    // Fallback to old structure
    if (service?.chemicalsReading) {
      return [{
        id: 'legacy',
        name: 'Chemical Readings',
        description: 'Legacy chemical readings',
        readings: Object.entries(service.chemicalsReading).map(([key, value]) => ({
          readingDefinitionId: key,
          value: value,
          readingDefinition: {
            name: key.charAt(0).toUpperCase() + key.slice(1),
            unit: key === 'ph' ? 'pH' : 'ppm'
          }
        }))
      }];
    }
    
    return [];
  };

  const getConsumableGroups = () => {
    if (service?.consumablesSnapshot) {
      return service.consumablesSnapshot.map((group: any) => ({
        ...group,
        consumables: service?.consumables?.filter((consumable: any) => 
          group.consumableDefinitions.some((def: any) => def.id === consumable.consumableDefinitionId)
        ) || []
      }));
    }
    
    // Fallback to old structure
    if (service?.chemicalsSpent) {
      return [{
        id: 'legacy',
        name: 'Chemicals Spent',
        description: 'Legacy chemicals spent',
        consumables: Object.entries(service.chemicalsSpent).map(([key, value]) => ({
          consumableDefinitionId: key,
          value: value,
          consumableDefinition: {
            name: key.charAt(0).toUpperCase() + key.slice(1),
            unit: key === 'liquidChlorine' || key === 'muriaticAcid' ? 'gallon' : 
                  key === 'tablet' ? 'unit' : 'oz'
          }
        }))
      }];
    }
    
    return [];
  };

  const getChecklistItems = () => {
    // Use new structured checklist if available
    if (service?.checklistSnapshot && service?.customChecklist) {
      return service.checklistSnapshot.map((item: any) => ({
        ...item,
        completed: service.customChecklist![item.id] || false
      }));
    }
    
    // Fallback to old checklist structure
    if (service?.checklist) {
      return [
        { id: 'poolVacuumed', label: 'Pool Vacuumed', completed: service.checklist.poolVacuumed },
        { id: 'skimmerCleaned', label: 'Skimmer Cleaned', completed: service.checklist.skimmerCleaned },
        { id: 'tilesBrushed', label: 'Tiles Brushed', completed: service.checklist.tilesBrushed },
        { id: 'pumpBasketCleaned', label: 'Pump Basket Cleaned', completed: service.checklist.pumpBasketCleaned },
        { id: 'filterWashed', label: 'Filter Washed', completed: service.checklist.filterWashed },
        { id: 'filterChanged', label: 'Filter Changed', completed: service.checklist.filterChanged },
        { id: 'chemicalsAdjusted', label: 'Chemicals Adjusted', completed: service.checklist.chemicalsAdjusted }
      ];
    }
    
    return [];
  };

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <DialogContent className="max-h-[90vh] max-w-4xl overflow-y-auto p-6">
        <DialogHeader>
          <div className="flex items-center gap-2">
            <ListChecks className="h-6 w-6 text-[#364D9D]" />
            <DialogTitle className="text-xl font-semibold">Service Report</DialogTitle>
          </div>

          {service.status === 'Completed' && (
            <div className="mt-2 flex items-center gap-2 text-sm text-gray-500">
              <div className="flex-1">
                {format(new Date(service?.completedAt || new Date()), "EEEE, MMMM do 'at' h:mm a")}
                <span className="ml-1 font-medium">
                  by {service?.completedByUser?.firstName || ''} {service?.completedByUser?.lastName || ''}
                </span>
              </div>
            </div>
          )}

          <div className="mt-1 flex items-center gap-1 text-sm text-gray-500">
            <MapPin className="h-4 w-4" />
            <span>
              {pool?.address}, {pool?.city}, {pool?.state}, {pool?.zip}
            </span>
          </div>
        </DialogHeader>

        {service.status === 'Completed' ? (
          <>
            <Accordion type="single" collapsible defaultValue={hasPhotos() ? "photos" : "readings"} className="mt-4">
              {/* Service Photos Section - First and Default */}
              {hasPhotos() && (
                <AccordionItem value="photos">
                  <AccordionTrigger className="text-lg">
                    <div className="flex items-center gap-2">
                      <ImageIcon className="h-5 w-5 text-[#364D9D]" />
                      <span className="text-md">Service Photos</span>
                    </div>
                  </AccordionTrigger>
                  <AccordionContent>
                    {service?.photosSnapshot ? (
                      // New structured photos
                      <div className="space-y-4">
                        {getPhotoGroups().map((group: any, groupIndex: number) => (
                          <Card key={groupIndex}>
                            <CardHeader>
                              <CardTitle className="text-lg">{group.name}</CardTitle>
                              {group.description && (
                                <p className="text-sm text-gray-600">{group.description}</p>
                              )}
                            </CardHeader>
                            <CardContent>
                              {group.photos.length > 0 ? (
                                <div className="grid grid-cols-1 gap-4 sm:grid-cols-2 md:grid-cols-3">
                                  {group.photos.map((photo: any, photoIndex: number) => {
                                    const photoDefinition = group.photoDefinitions.find((def: any) => def.id === photo.photoDefinitionId);
                                    return (
                                      <div key={photoIndex} className="space-y-2">
                                        <div className="relative aspect-square overflow-hidden rounded-lg">
                                          <Image
                                            src={photo.url}
                                            alt={photoDefinition?.name || `Service photo ${photoIndex + 1}`}
                                            layout="fill"
                                            className="object-cover transition-transform hover:scale-105"
                                          />
                                        </div>
                                        {photoDefinition && (
                                          <p className="text-sm font-medium text-gray-700">{photoDefinition.name}</p>
                                        )}
                                        {photo.notes && (
                                          <p className="text-xs text-gray-500">{photo.notes}</p>
                                        )}
                                      </div>
                                    );
                                  })}
                                </div>
                              ) : (
                                <p className="text-sm text-gray-500">No photos available for this group.</p>
                              )}
                            </CardContent>
                          </Card>
                        ))}
                      </div>
                    ) : (
                      // Legacy photos
                      <div className="grid grid-cols-1 gap-4 sm:grid-cols-2 md:grid-cols-3">
                        {getStructuredPhotos()?.map((photo: any, index: number) => (
                          <div key={index} className="relative aspect-square overflow-hidden rounded-lg">
                            <Image
                              src={photo.url}
                              alt={`Service photo ${index + 1}`}
                              layout="fill"
                              className="object-cover transition-transform hover:scale-105"
                            />
                          </div>
                        ))}
                      </div>
                    )}
                  </AccordionContent>
                </AccordionItem>
              )}

              {/* Readings Section */}
              {getReadingGroups().length > 0 && (
                <AccordionItem value="readings">
                  <AccordionTrigger className="text-lg">
                    <div className="flex items-center gap-2">
                      <Beaker className="h-5 w-5 text-[#364D9D]" />
                      <span className="text-md">Readings</span>
                    </div>
                  </AccordionTrigger>
                  <AccordionContent>
                    <div className="space-y-4">
                      {getReadingGroups().map((group: any, groupIndex: number) => (
                        <Card key={groupIndex}>
                          <CardHeader>
                            <CardTitle className="text-lg">{group.name}</CardTitle>
                            {group.description && (
                              <p className="text-sm text-gray-600">{group.description}</p>
                            )}
                          </CardHeader>
                          <CardContent className="grid grid-cols-2 gap-4 sm:grid-cols-3">
                            {group.readings.length > 0 ? (
                              group.readings.map((reading: any, readingIndex: number) => (
                                <div key={readingIndex} className="rounded-lg bg-gray-50 p-3">
                                  <div className="text-sm text-gray-500">
                                    {reading.readingDefinition?.name || reading.readingDefinitionId} 
                                    {reading.readingDefinition?.unit && ` (${reading.readingDefinition.unit})`}
                                  </div>
                                  <div className="text-lg font-semibold">{reading.value || 'N/A'}</div>
                                </div>
                              ))
                            ) : (
                              // Fallback for legacy structure
                              group.readingDefinitions?.map((def: any, defIndex: number) => {
                                const reading = group.readings.find((r: any) => r.readingDefinitionId === def.id);
                                return (
                                  <div key={defIndex} className="rounded-lg bg-gray-50 p-3">
                                    <div className="text-sm text-gray-500">
                                      {def.name} {def.unit && `(${def.unit})`}
                                    </div>
                                    <div className="text-lg font-semibold">{reading?.value || 'N/A'}</div>
                                  </div>
                                );
                              })
                            )}
                          </CardContent>
                        </Card>
                      ))}
                    </div>
                  </AccordionContent>
                </AccordionItem>
              )}

              {/* Consumables Section */}
              {getConsumableGroups().length > 0 && (
                <AccordionItem value="consumables">
                  <AccordionTrigger className="text-lg">
                    <div className="flex items-center gap-2">
                      <FlaskConical className="h-5 w-5 text-[#364D9D]" />
                      <span className="text-md">Consumables</span>
                    </div>
                  </AccordionTrigger>
                  <AccordionContent>
                    <div className="space-y-4">
                      {getConsumableGroups().map((group: any, groupIndex: number) => (
                        <Card key={groupIndex}>
                          <CardHeader>
                            <CardTitle className="text-lg">{group.name}</CardTitle>
                            {group.description && (
                              <p className="text-sm text-gray-600">{group.description}</p>
                            )}
                          </CardHeader>
                          <CardContent className="grid grid-cols-2 gap-4 sm:grid-cols-3">
                            {group.consumables.length > 0 ? (
                              group.consumables.map((consumable: any, consumableIndex: number) => (
                                <div key={consumableIndex} className="rounded-lg bg-gray-50 p-3">
                                  <div className="text-sm text-gray-500">
                                    {consumable.consumableDefinition?.name || consumable.consumableDefinitionId}
                                    {consumable.consumableDefinition?.unit && ` (${consumable.consumableDefinition.unit})`}
                                  </div>
                                  <div className="text-lg font-semibold">
                                    {consumable.value ? `${consumable.value} ${consumable.consumableDefinition?.unit || ''}` : 'N/A'}
                                  </div>
                                </div>
                              ))
                            ) : (
                              // Fallback for legacy structure
                              group.consumableDefinitions?.map((def: any, defIndex: number) => {
                                const consumable = group.consumables.find((c: any) => c.consumableDefinitionId === def.id);
                                return (
                                  <div key={defIndex} className="rounded-lg bg-gray-50 p-3">
                                    <div className="text-sm text-gray-500">
                                      {def.name} {def.unit && `(${def.unit})`}
                                    </div>
                                    <div className="text-lg font-semibold">
                                      {consumable?.value ? `${consumable.value} ${def.unit || ''}` : 'N/A'}
                                    </div>
                                  </div>
                                );
                              })
                            )}
                          </CardContent>
                        </Card>
                      ))}
                    </div>
                  </AccordionContent>
                </AccordionItem>
              )}

              {/* Checklist Section */}
              {getChecklistItems().length > 0 && (
                <AccordionItem value="checklist">
                  <AccordionTrigger className="text-lg">
                    <div className="flex items-center gap-2">
                      <ListChecks className="h-5 w-5 text-[#364D9D]" />
                      <span className="text-md">Service Checklist</span>
                    </div>
                  </AccordionTrigger>
                  <AccordionContent>
                    <Card>
                      <CardContent className="grid grid-cols-1 gap-2 p-4 sm:grid-cols-2">
                        {getChecklistItems().map((item: any, index: number) => (
                          <div key={index} className="flex items-center justify-between rounded-lg bg-gray-50 p-3">
                            <span className="text-gray-700">{item.label || item.name}</span>
                            {item.completed ? (
                              <CheckCircleIcon className="h-5 w-5 text-green-600" />
                            ) : (
                              <XCircleIcon className="h-5 w-5 text-red-600" />
                            )}
                          </div>
                        ))}
                      </CardContent>
                    </Card>
                  </AccordionContent>
                </AccordionItem>
              )}

            </Accordion>

            {/* Actions */}
            <div className="mt-6 flex justify-center gap-3">
              <Button
                onClick={handleResendEmail}
                variant="outline"
                className="flex items-center gap-2"
                disabled={resendEmailMutation.isPending}
              >
                <Mail className="h-4 w-4" />
                {resendEmailMutation.isPending ? 'Sending...' : 'Resend Email'}
              </Button>

              {hasPhotos() && (
                <Button
                  onClick={async () => {
                    try {
                      const photosForDownload = getPhotosForDownload();
                      const zipContent = await zipImages(photosForDownload);
                      const url = URL.createObjectURL(zipContent);
                      const link = document.createElement('a');
                      link.href = url;
                      link.download = 'service-photos.zip';
                      link.click();
                      URL.revokeObjectURL(url);
                    } catch (error) {
                      console.error('Error downloading photos:', error);
                      alert('Failed to download photos. Please try again.');
                    }
                  }}
                  variant="outline"
                  className="flex items-center gap-2"
                >
                  <Download className="h-4 w-4" />
                  Download Photos
                </Button>
              )}

              <DeleteServiceDialog
                serviceId={service.id}
                assignmentId={service.assignmentId}
                clientId={pool.clientOwnerId}
                onSuccess={() => setOpen(false)}
                trigger={
                  <Button variant="destructive" className="flex items-center gap-2">
                    <Trash2 className="h-4 w-4" />
                    Delete Service
                  </Button>
                }
              />
            </div>
          </>
        ) : (
          <div className="flex h-40 items-center justify-center text-lg text-gray-500">
            {getStatusMessage(service.status)}
          </div>
        )}
      </DialogContent>
    </Dialog>
  );
}
