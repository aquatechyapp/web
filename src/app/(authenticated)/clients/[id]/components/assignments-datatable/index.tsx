import { BasicDataTable } from '@/components/basic-datatable';
import { columns, PoolAssignmentsPopulated } from './columns';
import { Assignment } from '@/ts/interfaces/Assignments';
import { clientAxios } from '@/lib/clientAxios';
import { useQuery } from '@tanstack/react-query';

export type AssignmentsDatatableProps = {
  data: Assignment[];
};

const getUserByIdFn = async (userId: string) => {
  const response = await clientAxios.get(`/users/${userId}`);
  return response.data.data.user;
};

const getPopulatedData = async (data: Assignment[]) =>
  await Promise.all(
    data.map(async (assignment) => {
      const user = await getUserByIdFn(assignment.assignmentToId);
      return {
        ...assignment,
        assignmentToUser: user
      };
    })
  );

export default function AssignmentsDatatable({ data }: AssignmentsDatatableProps) {
  const { data: populatedData } = useQuery<PoolAssignmentsPopulated[]>({
    queryKey: ['assignments-populated'],
    queryFn: () => getPopulatedData(data)
  });

  return <BasicDataTable columns={columns} data={populatedData || []} />;
}
