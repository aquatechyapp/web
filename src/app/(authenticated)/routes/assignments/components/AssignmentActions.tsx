import { useState } from 'react';
import { BsThreeDotsVertical } from 'react-icons/bs';
import { Button } from '@/components/ui/button';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger
} from '@/components/ui/dropdown-menu';
import { Assignment } from '@/ts/interfaces/Assignments';
import { DialogDeleteAssignment } from '../ModalDeleteAssignment';
import { DialogTransferRoute } from '../ModalTransferRoute';

interface AssignmentActionsProps {
  assignment: Assignment;
}

export function AssignmentDropdownActions({ assignment }: AssignmentActionsProps) {
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [isTransferModalOpen, setIsTransferModalOpen] = useState(false);
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);

  const handleActionSelect = (action: 'transfer' | 'delete') => {
    if (action === 'transfer') {
      setIsTransferModalOpen(true);
    } else if (action === 'delete') {
      setIsDeleteModalOpen(true);
    }
    setIsDropdownOpen(false);
  };

  return (
    <>
      <DropdownMenu open={isDropdownOpen} onOpenChange={setIsDropdownOpen}>
        <DropdownMenuTrigger asChild className="self-center">
          <Button size="icon" variant="ghost">
            <BsThreeDotsVertical className="text-stone-500" />
          </Button>
        </DropdownMenuTrigger>
        <DropdownMenuContent align="end">
          <DropdownMenuItem onSelect={() => handleActionSelect('transfer')}>Transfer Route</DropdownMenuItem>
          <DropdownMenuItem onSelect={() => handleActionSelect('delete')} className="text-red-500">
            Delete Assignment
          </DropdownMenuItem>
        </DropdownMenuContent>
      </DropdownMenu>

      <DialogDeleteAssignment open={isDeleteModalOpen} setOpen={setIsDeleteModalOpen} assignment={assignment} />

      <DialogTransferRoute open={isTransferModalOpen} setOpen={setIsTransferModalOpen} assignment={assignment} />
    </>
  );
}
