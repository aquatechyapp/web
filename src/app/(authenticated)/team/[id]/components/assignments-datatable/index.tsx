import { BasicDataTable } from '@/components/basic-datatable';
import { columns, PoolAssignmentsPopulated } from './columns';
import { Assignment } from '@/ts/interfaces/Assignments';
import { clientAxios } from '@/lib/clientAxios';
import { useQuery } from '@tanstack/react-query';
import { useEffect } from 'react';

export type AssignmentsDatatableProps = {
  data: Assignment[];
  onLoadingComplete?: () => void;
};

const getUserByIdFn = async (userId: string) => {
  const response = await clientAxios.get(`/users/${userId}`);
  return response.data.data.user;
};

const getPopulatedData = async (data: Assignment[]) => {
  if (!data || data.length === 0) return [];

  return await Promise.all(
    data.map(async (assignment) => {
      const user = await getUserByIdFn(assignment.assignmentToId);
      return {
        ...assignment,
        assignmentToUser: user
      };
    })
  );
};

export default function AssignmentsDatatable({ data, onLoadingComplete }: AssignmentsDatatableProps) {
  const {
    data: populatedData,
    isLoading,
    isSuccess,
    isError
  } = useQuery<PoolAssignmentsPopulated[]>({
    queryKey: ['assignments-populated', data.map((a) => a.id)],
    queryFn: () => getPopulatedData(data),
    enabled: true,
    retry: false,
    staleTime: 30000
  });

  useEffect(() => {
    if (!isLoading) {
      onLoadingComplete?.();
    }
  }, [isLoading, onLoadingComplete]);

  return <BasicDataTable columns={columns} data={populatedData || []} isLoading={isLoading} />;
}
