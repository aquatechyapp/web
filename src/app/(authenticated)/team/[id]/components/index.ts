export { EmailPreferencesCard } from './EmailPreferencesCard';
export { FilterMaintenanceCard } from './FilterMaintenanceCard';
export { CustomChecklistCard } from './CustomChecklistCard';
export { ChecklistTemplatesCard } from './ChecklistTemplatesCard';
export { ReadingAndConsumableGroupsCard } from './ReadingAndConsumableGroupsCard';
export { ServiceTypesCard } from './ServiceTypesCard';

// Reading Groups
export * from './ReadingAndConsumableGroupsCard/ReadingGroups';

// Consumable Groups
export * from './ReadingAndConsumableGroupsCard/ConsumableGroups';

// Photo Groups
export * from './ReadingAndConsumableGroupsCard/PhotoGroups';

// Selector Groups
export * from './ReadingAndConsumableGroupsCard/SelectorGroups';

// Checklist Templates
export * from './ChecklistTemplatesCard';

// Service Types
export * from './ServiceTypesCard';
