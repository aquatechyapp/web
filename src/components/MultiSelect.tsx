import { Check, ChevronsUpDown, X } from 'lucide-react';
import * as React from 'react';

import { Badge } from '@/components/ui/badge';
import { Button } from '@/components/ui/button';
import { Command, CommandEmpty, CommandGroup, CommandInput, CommandItem } from '@/components/ui/command';
import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/popover';
import { cn } from '@/lib/utils';

export type OptionType = {
  label: string;
  value: string;
};

interface MultiSelectProps {
  options: OptionType[];
  selected: string[];
  onChange: (selected: string[]) => void;
  className?: string;
  placeholder: string;
}

// Atualmente não será usado, mas futuramente provavelmente será

function MultiSelect({ options, selected, onChange, placeholder, ...props }: MultiSelectProps) {
  const [open, setOpen] = React.useState(false);

  const handleUnselect = (item: string) => {
    onChange(selected.filter((i) => i !== item));
  };

  return (
    <div className="w-[100%] ">
      <Popover open={open} onOpenChange={setOpen} {...props}>
        <PopoverTrigger asChild>
          <Button
            variant="outline"
            role="combobox"
            aria-expanded={open}
            className={`mt-2 w-full justify-between ${selected.length > 1 ? 'h-full' : 'h-9'}`}
            onClick={() => setOpen(!open)}
          >
            <div className="flex flex-wrap gap-1">
              {selected.map((item) => (
                <Badge variant="secondary" key={item} className="mb-1 mr-1" onClick={() => handleUnselect(item)}>
                  {item ? item : 'cityes'}
                  <button
                    className="ring-offset-background focus:ring-ring ml-1 rounded-full outline-none focus:ring-2 focus:ring-offset-2"
                    onKeyDown={(e) => {
                      if (e.key === 'Enter') {
                        handleUnselect(item);
                      }
                    }}
                    onMouseDown={(e) => {
                      e.preventDefault();
                      e.stopPropagation();
                    }}
                    onClick={() => handleUnselect(item)}
                  >
                    <X className="text-muted-foreground hover:text-foreground h-3 w-3" />
                  </button>
                </Badge>
              ))}
            </div>
            {!selected.length && <p className="w-full text-start text-slate-500">{placeholder}</p>}
            <ChevronsUpDown className="h-4 w-4 shrink-0 text-slate-500 opacity-50" />
          </Button>
        </PopoverTrigger>
        <PopoverContent align="start" className="p-0">
          <Command className="">
            <CommandInput placeholder="Search ..." />
            <CommandEmpty>No item found.</CommandEmpty>
            <CommandGroup className="max-h-46 overflow-auto">
              {options.map((option) => (
                <CommandItem
                  key={option.value}
                  onSelect={() => {
                    onChange(
                      selected.includes(option.value)
                        ? selected.filter((item) => item !== option.value)
                        : [...selected, option.value]
                    );
                    setOpen(true);
                  }}
                >
                  <Check
                    className={cn('mr-2 h-4 w-4', selected.includes(option.value) ? 'opacity-100' : 'opacity-0')}
                  />
                  {option.label}
                </CommandItem>
              ))}
            </CommandGroup>
          </Command>
        </PopoverContent>
      </Popover>
    </div>
  );
}

export { MultiSelect };
