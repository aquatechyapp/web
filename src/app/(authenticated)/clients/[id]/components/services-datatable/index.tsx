import { useState } from 'react';
import { BasicDataTable } from '@/components/basic-datatable';
import { columns } from './columns';
import { Service } from '@/ts/interfaces/Service';
import { Pool } from '@/ts/interfaces/Pool';
import { ModalViewService } from '../ModalViewService';
import { BasicServicesDataTable } from '@/components/basic-services-datatable';

export type ServicesDatatableProps = {
  services: Service[];
  pool: Pool;
};

export default function ServicesDatatable({ services, pool }: ServicesDatatableProps) {
  const [selectedService, setSelectedService] = useState<Service | null>(null);
  const [open, setOpen] = useState(false);

  // Sort services by scheduledTo date (most recent first)
  const sortedServices = [...services].sort((a, b) => {
    const dateA = new Date(a.scheduledTo);
    const dateB = new Date(b.scheduledTo);
    return dateB.getTime() - dateA.getTime();
  });

  return (
    <>
      <BasicServicesDataTable
        columns={columns}
        data={sortedServices}
        onRowClick={(row) => {
          setSelectedService(row.original);
          setOpen(true);
        }}
      />

      {selectedService && <ModalViewService service={selectedService} pool={pool} open={open} setOpen={setOpen} />}
    </>
  );
}
