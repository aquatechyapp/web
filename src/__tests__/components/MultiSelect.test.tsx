import { render } from '@testing-library/react';

import { MultiSelect } from '@/components/MultiSelect';

describe('MultiSelect', () => {
  it('should render the MultiSelect component', () => {
    const wrapper = render(
      <MultiSelect
        options={[{ label: 'label', value: 'value' }]}
        selected={['selected']}
        onChange={() => {}}
        placeholder="placeholder"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
