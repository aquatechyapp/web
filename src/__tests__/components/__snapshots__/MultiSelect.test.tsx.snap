// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiSelect > should render the MultiSelect component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-[100%]"
      >
        <button
          aria-controls="radix-:r0:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="inline-flex items-center gray-50space-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-slate-300 border border-slate-200 bg-gray-50 shadow-sm hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:bg-slate-950 dark:hover:bg-slate-800 dark:hover:text-slate-50 px-4 py-2 mt-2 w-full justify-between h-9"
          data-state="closed"
          role="combobox"
          type="button"
        >
          <div
            class="flex flex-wrap gap-1"
          >
            <div
              class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-slate-950 focus:ring-offset-2 dark:border-slate-800 dark:focus:ring-slate-300 border-transparent bg-slate-100 text-slate-900 hover:bg-slate-100/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80 mb-1 mr-1"
            >
              selected
              <button
                class="ring-offset-background focus:ring-ring ml-1 rounded-full outline-none focus:ring-2 focus:ring-offset-2"
              >
                <svg
                  class="lucide lucide-x text-muted-foreground hover:text-foreground h-3 w-3"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 6 6 18"
                  />
                  <path
                    d="m6 6 12 12"
                  />
                </svg>
              </button>
            </div>
          </div>
          <svg
            class="lucide lucide-chevrons-up-down h-4 w-4 shrink-0 text-slate-500 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m7 15 5 5 5-5"
            />
            <path
              d="m7 9 5-5 5 5"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-[100%]"
    >
      <button
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center gray-50space-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-slate-300 border border-slate-200 bg-gray-50 shadow-sm hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:bg-slate-950 dark:hover:bg-slate-800 dark:hover:text-slate-50 px-4 py-2 mt-2 w-full justify-between h-9"
        data-state="closed"
        role="combobox"
        type="button"
      >
        <div
          class="flex flex-wrap gap-1"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-slate-950 focus:ring-offset-2 dark:border-slate-800 dark:focus:ring-slate-300 border-transparent bg-slate-100 text-slate-900 hover:bg-slate-100/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80 mb-1 mr-1"
          >
            selected
            <button
              class="ring-offset-background focus:ring-ring ml-1 rounded-full outline-none focus:ring-2 focus:ring-offset-2"
            >
              <svg
                class="lucide lucide-x text-muted-foreground hover:text-foreground h-3 w-3"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 6 6 18"
                />
                <path
                  d="m6 6 12 12"
                />
              </svg>
            </button>
          </div>
        </div>
        <svg
          class="lucide lucide-chevrons-up-down h-4 w-4 shrink-0 text-slate-500 opacity-50"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 15 5 5 5-5"
          />
          <path
            d="m7 9 5-5 5 5"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
